### PR TITLE
JDK-8319784: VM crash during heap dump after JDK-8287061

### PR DESCRIPTION
**Pull request by JohnTortugo**
This fixes an incorrect assert in `debugInfo.hpp`. The assert is incorrect because this method may be called before rematerialization of the ObjectValue takes place, and therefore, `_selected` will be nullptr. This pull request patches the method to just return `Handle() => nullptr` if the rematerialization of the object hasn't taken place yet. This is the same behavior used by scalar-replaced objects not participating in merges.

I added a test to reproduce a scenario where the assert was causing a problem.

Tested on Ubuntu/MacOS/Windows x86_64 tier1-4.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8319784](https://bugs.openjdk.org/browse/JDK-8319784): VM crash during heap dump after JDK-8287061 (**Bug** - P2)

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to [aa723f50](https://git.openjdk.org/jdk/pull/16622/files/aa723f50fca4f46e3d28aaccae4635ede0c8cf9e)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**) ⚠️ Review applies to [26f7c936](https://git.openjdk.org/jdk/pull/16622/files/26f7c9365565396999caa83e571b682816096d55)
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - **Reviewer**) ⚠️ Review applies to [26f7c936](https://git.openjdk.org/jdk/pull/16622/files/26f7c9365565396999caa83e571b682816096d55)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16622/head:pull/16622` \
`$ git checkout pull/16622`

Update a local copy of the PR: \
`$ git checkout pull/16622` \
`$ git pull https://git.openjdk.org/jdk.git pull/16622/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16622`

View PR using the GUI difftool: \
`$ git pr show -t 16622`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16622.diff">https://git.openjdk.org/jdk/pull/16622.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16622#issuecomment-1809347536)

### CODE REVIEW

**Review by vnkozlov — APPROVED** [review]
Looks good. You need second review.

File: test/hotspot/jtreg/compiler/c2/TestReduceAllocationAndHeapDump.java
Comment (shipilev) [inline test/hotspot/jtreg/compiler/c2/TestReduceAllocationAndHeapDump.java:32]: I am concerned this test case would deplete the disk space with heap dumps. I think the test should explicitly trigger the heap dump and manage (remove) the resulting .hprofs.

Comment (TobiHartmann) [reply test/hotspot/jtreg/compiler/c2/TestReduceAllocationAndHeapDump.java:32]: I agree, especially if the test is executed with stress options like `FullGCALot`.

Comment (JohnTortugo) [reply test/hotspot/jtreg/compiler/c2/TestReduceAllocationAndHeapDump.java]: Thank you for reviewing guys! I pushed changes to the test to address your feedback.

**Review by TobiHartmann — APPROVED** [review]
Looks good to me. I've re-submitted testing and will report back once it passed.

**Comment by TobiHartmann** [comment]
All tests passed.

File: test/hotspot/jtreg/compiler/c2/TestReduceAllocationAndHeapDump.java
Comment (shipilev) [inline test/hotspot/jtreg/compiler/c2/TestReduceAllocationAndHeapDump.java:47]: Java style: `dumperArgs`

File: test/hotspot/jtreg/compiler/c2/TestReduceAllocationAndHeapDump.java
Comment (shipilev) [inline test/hotspot/jtreg/compiler/c2/TestReduceAllocationAndHeapDump.java:79]:
```suggestion
        for (int i = 0; i < 5000; i++) {
```

**Review by shipilev — APPROVED** [review]
Good. I have just a few cosmetic suggestions.

**Pull request closed without merging**
